lib/SIL/SILVerifier.cpp

```diff
@@ -4847,6 +4851,54 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
     }
   }
 
+  void verifyLinetable(SILBasicBlock *BB) {
+    if (!VerifyLinetableMonotonicity)
+      return;
+
+    SILFunction *F = BB->getParent();
+
+    // This check only makes sense at -Onone. Optimizations, e.g. inlining,
+    // can move instructions around violate the monotonicity.
+    if (F->getEffectiveOptimizationMode() != OptimizationMode::NoOptimization)
+      return;
+
+    // Transparent functions don't have debug info.
+    if (F->isTransparent())
+      return;
+
+    auto &SrcMgr = BB->getParent()->getModule().getSourceManager();
+    SILDebugLocation LastLoc;
+
+    using UUSTuple = std::pair<std::pair<unsigned, unsigned>, StringRef>;
+    struct DebugLocKey : public UUSTuple {
+      DebugLocKey(SILLocation::DebugLoc DL)
+          : UUSTuple({{DL.Line, DL.Column}, DL.Filename}) {}
+      inline bool operator==(const SILLocation::DebugLoc &DL) const {
+        return first.first == DL.Line && first.second == DL.Column &&
+               second.equals(DL.Filename);
+      }
+    };
+    llvm::DenseSet<UUSTuple> VisitedLocations;
+
+    for (SILInstruction &SI : *BB) {
+      if (SI.isMetaInstruction())
+        continue;
+      if (isa<DeallocStackInst>(&SI))
+        continue;
+      if (isa<MetatypeInst>(&SI))
+        continue;
+      auto LocKind = SI.getDebugLocation().getLocation().getKind();
+      if (LocKind != SILLocation::LocationKind::RegularKind)
+        continue;
+      auto DL = SI.getDebugLocation().getLocation().decodeDebugLoc(SrcMgr);
+      if (DL.Line == 0)
+        continue;
+      auto ItNew = DebugLocKey(DL);
+      require(!VisitedLocations.count(ItNew), "Incorrect location at -Onone");
```

Comment:
I don't think we make any promise that two different AST expressions can't have the same location, and we certainly don't promise that any particular AST expression only corresponds to one SIL instruction. What are you actually trying to test?

Reply:
The idea we came up with Adrian is that of trying to add a pass to make sure we don't have locations jumping back within a single basic block.
The motivation is that, even at `-Onone`, we see sometimes stepping in the debugger jumping back in straight line code, and we thought we could add a verifier similarly to what we did for debug scopes [which helped catching a lot of bugs].

To give you some more context, the verifier was originally written in `IRGenDebugInfo.cpp`, and there are still some vestiges of it there (`-verify-linetable`), then we decided to move it at SIL level, so we can catch which optimization breaks the invariants running `-sil-verify-all`.

The main invariant that the current code tries to maintain is making sure that we don't end up bouncing back to an existing location, but from what you say this is something we shouldn't rely on? Do you think there's a better way of achieving what we're trying to do here?

Thanks!

Reply:
I think it'd be reasonable to say that you'd never go backwards to a location in a previous *statement,* as well as making a policy of not revisiting the same AST *expression.* But I don't think you can do the latter based on source locations, and the former requires using a ParentMap or something to know which locations are within which statements.

Reply:
Even that gets tricky with inlined closures.